14/Add polling and waiting instead of naive wait for TalkBack toggling

ataulm opened the pull request:
Fixes #14 

We added the 1500ms wait because _occasionally_ the tests would fail because they were switching Talkback on/off too fast, and it hadn't completed when the test started.

If you have a large suite of tests, this isn't great, because it adds 3s per test.

Swapped that with a while loop and querying to see if TalkBack is in the desired state. Ran the suite 10 times and it didn't fail.

ataulm committed:
Add polling and waiting instead of naive wait for TalkBack toggling

xrigau commented on core/src/main/java/com/novoda/espresso/TalkBackTestRule.java:
```diff
@@ -27,16 +33,32 @@ public Statement apply(Statement base, Description description) {
+        }
+
+        private void sleepUntilTalkBackIsEnabled() {
+            while (talkBackDisabled()) {
```
maybe add a timeout? otherwise it could potentially sleep forever

xrigau replied:
i.e. stop after 150 iterations regardless of the state of talkback

ataulm replied:
I'm not sure this is true - I will check by adding a 30min sleep and watching a movie.

ataulm replied:
commit 3d3e7c5e78772c1e9e8192ad0772ae9f58858e87
Author: Ataul Munim <hello@ataulm.com>
Date:   Fri Jan 12 15:14:39 2018 +0000

    Add limit to retries for waiting for TalkBack toggle until test fails


https://github.com/novoda/espresso-support/pull/15#issuecomment-357265279

ataulm committed:
Add limit to retries for waiting for TalkBack toggle until test fails

ataulm commented:
@xrigau You're right, even if the test runner will eventually timeout, we should fail earlier. Since 1500ms was stable before, I chose 100ms delay and 15 retries to begin with, which in the best case should be 15x faster for each test.

The assertion thrown is test specific too - it will try to go onto the next test if one fails because of the timeout.

xrigau commented on core/src/main/java/com/novoda/espresso/TalkBackTestRule.java:
```diff
@@ -27,16 +36,49 @@ public Statement apply(Statement base, Description description) {
+            return new Condition() {
+                @Override
+                public boolean holds() {
+                    return !talkBackIsDisabled().holds();
```
this is test land but creating a `Condition` instance every time the `holds` method is called seems a bit unnecessary?

ataulm replied:
:+1:

commit afa7a6639a355262316a1b6801637dd80c8e17de
Author: Ataul Munim <hello@ataulm.com>
Date:   Fri Jan 12 16:48:13 2018 +0000

    Extract check for TalkBack
    
    - this method can be used in each condition, rather than instantiating a new disabledCondition
    each time the enabledCondition.holds() method is called

ataulm committed:
Extract check for TalkBack

- this method can be used in each condition, rather than instantiating a new disabledCondition
each time the enabledCondition.holds() method is called

xrigau reviewed (approved):
👍 noice!

ataulm commented:
thanks para the thoughtful review! 🙏
